### PR TITLE
修改BufferPoolMetricSet支持JMX采集远程指标

### PR DESCRIPTION
**Pull request by RobertoHuang**
修改BufferPoolMetricSet支持JMX采集远程指标

修改MetricRegistryImpl添加取消所有已注册指标的方法

### CODE REVIEW

**Comment by ralf0131** [comment]
Could you config your email in the commit to be the same as your Github account?
Your pull request will not be merged until you have signed  CLA.

File: metrics-core-impl/src/main/java/com/alibaba/metrics/MetricRegistryImpl.java [hunk 1]
```diff
@@ -65,6 +65,20 @@ public MetricRegistryImpl(int maxMetricCount) {
...
     }
 
+    /**
+     * 〈unRegister all metrics.〉
+     *
+     *  eg:
+     *      When collecting remote machine JVM metrics through the JMX port, May be due to network jitter or
+     *      remote process restart, Need to cancel the previously collected metrics and re-register.
+     *
+     * @author HuangTaiHong
```
Comment (ralf0131) [inline metrics-core-impl/src/main/java/com/alibaba/metrics/MetricRegistryImpl.java]: Please remove author name.

File: metrics-openjdk/src/main/java/com/alibaba/metrics/jvm/BufferPoolMetricSet.java [hunk 2]
```diff
@@ -42,10 +43,10 @@
     private static final String[] NAMES = { "count", "used", "capacity" };
     private static final String[] POOLS = { "direct", "mapped" };
 
-    private final MBeanServer mBeanServer;
+    private final MBeanServerConnection mBeanServerConnection;
 
-    public BufferPoolMetricSet(MBeanServer mBeanServer) {
-        this.mBeanServer = mBeanServer;
+    public BufferPoolMetricSet(MBeanServerConnection mBeanServerConnection) {
```
Comment (ralf0131) [inline metrics-openjdk/src/main/java/com/alibaba/metrics/jvm/BufferPoolMetricSet.java]: Please keep the original constructor and add a new one.

**Comment by ralf0131** [comment]
Please make sure the code can be compiled on Java 6.

```
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/travis/build/alibaba/metrics/metrics-openjdk/src/test/java/com/alibaba/metrics/jvm/JmxRemoteMetricsTest.java:[37,54] error: diamond operator is not supported in -source 6
[INFO] 1 error
```

File: metrics-openjdk/src/main/java/com/alibaba/metrics/jvm/BufferPoolMetricSet.java (hunk 2 above)
Comment (RobertoHuang) [reply metrics-openjdk/src/main/java/com/alibaba/metrics/jvm/BufferPoolMetricSet.java]: MBeanServer是MBeanServerConnection的子类
因为后续没有用到MBeanServer的特性 这样修改原有的代码也不会出错
所以我觉得原来的构造不需要保留

File: metrics-core-impl/src/main/java/com/alibaba/metrics/MetricRegistryImpl.java (hunk 1 above)
Comment (RobertoHuang) [reply metrics-core-impl/src/main/java/com/alibaba/metrics/MetricRegistryImpl.java]: 已修改

File: metrics-openjdk/src/main/java/com/alibaba/metrics/jvm/BufferPoolMetricSet.java (hunk 2 above)
Comment (ralf0131) [reply metrics-openjdk/src/main/java/com/alibaba/metrics/jvm/BufferPoolMetricSet.java]: +1 Look good.

File: metrics-openjdk/src/test/java/com/alibaba/metrics/jvm/JmxRemoteMetricsTest.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/**
+ * FileName: JmxRemoteMetricsTest
```
Comment (ralf0131) [inline metrics-openjdk/src/test/java/com/alibaba/metrics/jvm/JmxRemoteMetricsTest.java]: Please remove the author name.

File: metrics-openjdk/src/test/java/com/alibaba/metrics/jvm/JmxRemoteMetricsTest.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import javax.management.remote.JMXConnectorFactory;
+import javax.management.remote.JMXServiceURL;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * 〈jmx remote metrics test.〉
+ *
+ * @author HuangTaiHong
```
Comment (ralf0131) [inline metrics-openjdk/src/test/java/com/alibaba/metrics/jvm/JmxRemoteMetricsTest.java]: Please remove the author name.

File: metrics-openjdk/src/test/java/com/alibaba/metrics/jvm/JmxRemoteMetricsTest.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    // @Test
+    public void testCollectionRemoteMetrics() throws IOException {
+        Map<String, Object> environment = new HashMap<String, Object>();
+        environment.put(JMXConnector.CREDENTIALS, new String[]{USERNAME, PASSWORD});
+        JMXServiceURL jmxServiceURL = new JMXServiceURL("service:jmx:rmi:///jndi/rmi://localhost:1099/jmxrmi");
+        MBeanServerConnection mBeanServerConnection = JMXConnectorFactory.connect(jmxServiceURL, environment).getMBeanServerConnection();
+        BufferPoolMetricSet bufferPoolMetricSet = new BufferPoolMetricSet(mBeanServerConnection);
+        for (Map.Entry<MetricName, Metric> entry : bufferPoolMetricSet.getMetrics().entrySet()) {
+            JmxAttributeGauge value = (JmxAttributeGauge) entry.getValue();
+            System.out.println(entry.getKey().getKey() + ":" + value.getValue());
```
Comment (ralf0131) [inline metrics-openjdk/src/test/java/com/alibaba/metrics/jvm/JmxRemoteMetricsTest.java]: Please mock the data and using assertion to check the expected value and actual value.

**Comment by ralf0131** [comment]
You have not signed the CLA, your pull request will not be merged until you have signed the CLA.

**Comment by RobertoHuang** [comment]
done. 辛苦啦!

**Pull request merged**